awx/main/tasks.py

```diff
@@ -149,6 +149,85 @@ def handle_setting_changes(self, setting_keys):
             break
 
 
+def _cancel_queues(app, worker_queues, worker_name):
+    removed_queues = []
+    for queue in worker_queues:
+        if queue['name'] in settings.AWX_CELERY_QUEUES_STATIC or \
+                {'alias': queue['name']} in settings.AWX_CELERY_QUEUES_STATIC:
+            continue
+
+        removed_queues.append(queue['name'])
+        app.control.cancel_consumer(queue['name'], reply=True, destination=[worker_name])
+    return removed_queues
+
+
+def _add_queues(app, instance, worker_name):
+    # Add worker to group-named queues
+    queues = []
+    for ig in instance.rampart_groups.all():
+        app.control.add_consumer(ig.name, reply=True, destination=[worker_name])
+        queues.append(ig.name)
+
+    # Add worker to worker-named queue
+    app.control.add_consumer(instance.hostname, reply=True, destination=[worker_name])
+    queues.append(instance.hostname)
+
+    return queues
+
+
+def _update_celery_routes(instance):
+    logger.debug("Updating celery routes for instance {}".format(instance.hostname))
+
+    tasks = [
+        'awx.main.tasks.cluster_node_heartbeat',
+        'awx.main.tasks.purge_old_stdout_files',
+    ]
+
+    # Instance is, effectively, a controller node
+    if instance.rampart_groups.filter(controller__isnull=False).exists():
+        tasks.append('awx.main.tasks.awx_isolated_heartbeat')
+
+    for t in tasks:
+        settings.CELERY_ROUTES[t] = {'queue': instance.hostname, 'routing_key': instance.hostname}
```

Comment:
so can we assume that this is thread-safe and all that jazz?

And this queue (based on instance.hostname, same as settings.CLUSTER_HOST_ID), I assume is fully local? Not shared across the cluster?

I see where it is added to `CELERY_QUEUES`, but only for the open source installer, and I can't really say where the actual `Queue` object is instantiated either in the development environment. I can kind of see where it is instantiated in production deploys, but it's murky.